lib/SILOptimizer/IPO/GlobalOpt.cpp

```diff
@@ -808,6 +857,50 @@ static bool canBeChangedExternally(SILGlobalVariable *SILG) {
   return true;
 }
 
+static bool canBeUsedOrChangedExternally(SILGlobalVariable *global) {
+  if (global->isLet())
+    return isPossiblyUsedExternally(global->getLinkage(),
+                                    global->getModule().isWholeModule());
+  return canBeChangedExternally(global);
+}
+
+static bool isSafeToRemove(SILGlobalVariable *global) {
+  return global->getDecl() && !canBeUsedOrChangedExternally(global);
+}
+
+/// If there are no loads or accesses of a given global, then remove its
+/// associated global addr and all asssociated instructions.
+bool SILGlobalOpt::tryRemoveGlobalAddr(SILGlobalVariable *global) {
+  if (!isSafeToRemove(global)) return false;
+
+  if (GlobalVarSkipProcessing.count(global) || GlobalLoadMap[global].size() ||
+      GlobalAccessMap[global].size())
+    return false;
+
+  for (auto *addr : GlobalAddrMap[global]) {
+    for (auto *use : addr->getUses()) {
+      if (!isa<StoreInst>(use->getUser()))
+        return false;
+    }
+    collectUsesOfInstructionForDeletion(addr);
+    InstToRemove.push_back(addr);
+  }
+
+  return true;
+}
+
+bool SILGlobalOpt::tryRemoveUnusedGlobal(SILGlobalVariable *global) {
+  if (!isSafeToRemove(global)) return false;
+
+  if (GlobalVarSkipProcessing.count(global) || GlobalAddrMap[global].size() ||
+      GlobalAccessMap[global].size() || GlobalLoadMap[global].size() ||
+      AllocGlobalStore.count(global) || GlobalVarStore.count(global))
+    return false;
+
+  Module->eraseGlobalVariable(global);
```

Comment:
Can this add SILGlobalVariables to delete to a vector as well? This will prevent having to create a vector that stores _all_ global variables...